File: specs/typespec/employer/tags.go
package employer

import (
	"fmt"
	"regexp"

	"vetchium-api-server.typespec/common"
)

var tagIDPattern = regexp.MustCompile(`^[a-z]([a-z0-9-]*[a-z0-9])?$`)

const tagIDMaxLength = 64

const (
	errTagIDRequired      = "tag_id is required"
	errTagIDTooLong       = "tag_id must be at most 64 characters"
	errTagIDInvalidFormat = "tag_id must contain only lowercase letters, digits, and hyphens, and must not start or end with a hyphen"
)

// GetTagRequest is the request body for POST /employer/get-tag.
type GetTagRequest struct {
	TagID  string `json:"tag_id"`
	Locale string `json:"locale,omitempty"`
}

func (r GetTagRequest) Validate() []common.ValidationError {
	return validateTagID(r.TagID)
}

// FilterTagsRequest is the request body for POST /employer/filter-tags.
type FilterTagsRequest struct {
	Query         string `json:"query,omitempty"`
	PaginationKey string `json:"pagination_key,omitempty"`
	Locale        string `json:"locale,omitempty"`
}

func (r FilterTagsRequest) Validate() []common.ValidationError {
	return nil
}

// Tag is the response type for portal tag reads.
type Tag struct {
	TagID        string  `json:"tag_id"`
	DisplayName  string  `json:"display_name"`
	Description  *string `json:"description,omitempty"`
	SmallIconURL *string `json:"small_icon_url,omitempty"`
	LargeIconURL *string `json:"large_icon_url,omitempty"`
}

// FilterTagsResponse is the response for POST /employer/filter-tags.
type FilterTagsResponse struct {
	Tags          []Tag  `json:"tags"`
	PaginationKey string `json:"pagination_key,omitempty"`
}

func validateTagID(tagID string) []common.ValidationError {
	var errs []common.ValidationError
	if tagID == "" {
		errs = append(errs, common.NewValidationError("tag_id", fmt.Errorf(errTagIDRequired)))
		return errs
	}
	if len(tagID) > tagIDMaxLength {
		errs = append(errs, common.NewValidationError("tag_id", fmt.Errorf(errTagIDTooLong)))
		return errs
	}
	if !tagIDPattern.MatchString(tagID) {
		errs = append(errs, common.NewValidationError("tag_id", fmt.Errorf(errTagIDInvalidFormat)))
	}
	return errs
}

File: api-schema/org/tags.ts
import { type ValidationError, newValidationError } from "../common/common";

const TAG_ID_PATTERN = /^[a-z]([a-z0-9-]*[a-z0-9])?$/;
const TAG_ID_MAX_LENGTH = 64;

export const ERR_TAG_ID_REQUIRED = "tag_id is required";
export const ERR_TAG_ID_TOO_LONG = `tag_id must be at most ${TAG_ID_MAX_LENGTH} characters`;
export const ERR_TAG_ID_INVALID_FORMAT =
	"tag_id must contain only lowercase letters, digits, and hyphens, and must not start or end with a hyphen";

export type TagId = string;

export function validateTagId(tagId: TagId): ValidationError[] {
	const errs: ValidationError[] = [];
	if (!tagId) {
		errs.push(newValidationError("tag_id", ERR_TAG_ID_REQUIRED));
		return errs;
	}
	if (tagId.length > TAG_ID_MAX_LENGTH) {
		errs.push(newValidationError("tag_id", ERR_TAG_ID_TOO_LONG));
		return errs;
	}
	if (!TAG_ID_PATTERN.test(tagId)) {
		errs.push(newValidationError("tag_id", ERR_TAG_ID_INVALID_FORMAT));
	}
	return errs;
}

// GetTagRequest is the request body for POST /org/get-tag.
export interface GetTagRequest {
	tag_id: TagId;
	locale?: string;
}

export function validateGetTagRequest(
	request: GetTagRequest
): ValidationError[] {
	return validateTagId(request.tag_id);
}

// FilterTagsRequest is the request body for POST /org/filter-tags.
export interface FilterTagsRequest {
	query?: string;
	pagination_key?: string;
	locale?: string;
}

export function validateFilterTagsRequest(
	_request: FilterTagsRequest
): ValidationError[] {
	return [];
}

// Tag is the response type for portal tag reads.
export interface Tag {
	tag_id: TagId;
	display_name: string;
	description?: string;
	small_icon_url?: string;
	large_icon_url?: string;
}

// FilterTagsResponse is the response for POST /org/filter-tags.
export interface FilterTagsResponse {
	tags: Tag[];
	pagination_key?: string;
}
